// @emdgroup-liquid/liquid/dist/collection/components/ld-tooltip/ld-tooltip.js
import Tether from 'tether';
import { Component, Element, h, Host, Listen, Method, Prop, State, Watch, } from '@stencil/core';
import { getClassNames } from '../../utils/getClassNames';
let tooltipCount = 0;
/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part icon - Default icon when no trigger is supplied
 * @part trigger - Trigger button
 * @part popper - Popper element (can only be styled as long as tooltip is not initialized)
 */
export class LdTooltip {
  constructor() {
    this.idDescriber = `ld-tooltip-${++tooltipCount}`;
    /** Delay in ms until tooltip hides (only when trigger type is 'hover') */
    this.hideDelay = 0;
    /** Position of the tooltip relative to the trigger element (also affects the arrow position) */
    this.position = 'top center';
    /** Delay in ms until tooltip shows (only when trigger type is 'hover') */
    this.showDelay = 0;
    /** The rendered HTML tag for the tooltip trigger. */
    this.tag = 'button';
    /** Event type that triggers the tooltip */
    this.triggerType = 'hover';
    this.hasDefaultTrigger = true;
    this.visible = false;
    this.mapPositionToAttachment = (position) => {
      return {
        'bottom center': 'top center',
        'bottom left': 'top left',
        'bottom right': 'top right',
        'left bottom': 'bottom right',
        'left middle': 'middle right',
        'left top': 'top right',
        'right bottom': 'bottom left',
        'right middle': 'middle left',
        'right top': 'top left',
        'top center': 'bottom center',
        'top left': 'bottom left',
        'top right': 'bottom right',
      }[position];
    };
    this.mapPositionToTargetAttachment = (position) => {
      var _a;
      return ((_a = {
        'left bottom': 'bottom left',
        'left middle': 'middle left',
        'left top': 'top left',
        'right bottom': 'bottom right',
        'right middle': 'middle right',
        'right top': 'top right',
      }[position]) !== null && _a !== void 0 ? _a : position);
    };
    this.initTooltip = async () => {
      const attachment = this.mapPositionToAttachment(this.position);
      const targetAttachment = this.mapPositionToTargetAttachment(this.position);
      const tooltipContent = this.tooltipRef.querySelector('slot').assignedNodes();
      tooltipContent.forEach((node) => {
        this.tooltipRef.appendChild(node);
      });
      const customTetherOptions = typeof this.tetherOptions === 'string'
        ? JSON.parse(this.tetherOptions)
        : this.tetherOptions;
      const tetherOptions = Object.assign({ attachment, classPrefix: 'ld-tether', constraints: [
          {
            attachment: 'together',
            to: 'window',
          },
        ], element: this.tooltipRef, target: this.triggerRef, targetAttachment }, customTetherOptions);
      this.popper = new Tether(tetherOptions);
      // Fixes a tether positioning bug
      this.popper.enable();
      this.popper.enable();
      this.popper.enable();
      this.popper.enable();
      this.visible = true;
    };
    this.toggleTooltip = () => {
      if (this.popper == undefined) {
        return;
      }
      if (this.visible) {
        this.hideTooltip();
      }
      else {
        this.showTooltip();
      }
    };
    this.handleHideTrigger = () => {
      if (this.triggerType === 'click' || this.disabled) {
        return;
      }
      clearTimeout(this.delayTimeout);
      if (this.popper) {
        this.delayTimeout = setTimeout(() => {
          this.hideTooltip();
        }, this.hideDelay);
      }
    };
    this.handleShowTrigger = () => {
      if (this.triggerType === 'click' || this.disabled) {
        return;
      }
      clearTimeout(this.delayTimeout);
      if (this.popper === undefined) {
        this.delayTimeout = setTimeout(this.initTooltip, this.showDelay);
      }
      else {
        this.delayTimeout = setTimeout(this.showTooltip.bind(this), this.showDelay);
      }
    };
    this.handleToggleTrigger = () => {
      if (this.triggerType === 'hover' || this.disabled) {
        return;
      }
      if (this.popper === undefined) {
        this.initTooltip();
      }
      else {
        this.toggleTooltip();
      }
    };
  }
  updatePopper(newDisabled) {
    if (newDisabled) {
      this.hideTooltip();
    }
  }
  /** Hide tooltip */
  async hideTooltip() {
    var _a;
    clearTimeout(this.delayTimeout);
    (_a = this.popper) === null || _a === void 0 ? void 0 : _a.disable();
    this.visible = false;
  }
  /** Show tooltip */
  async showTooltip() {
    if (this.disabled)
      return;
    clearTimeout(this.delayTimeout);
    this.popper.enable();
    this.visible = true;
  }
  // TODO: maybe this should listen only, if the tooltip was opened by click.
  handleClickOutside(event) {
    if (this.popper &&
      this.triggerType === 'click' &&
      event.target.closest('ld-tooltip') !== this.element &&
      event.target.closest('.ld-tooltip') !== this.tooltipRef) {
      this.hideTooltip();
    }
  }
  // Mobile Safari in some cases does not react to click events on elements
  // which are not interactive. But it does to touch events.
  // TODO: maybe this should listen only, if the tooltip was opened by click.
  handleTouchOutside(event) {
    this.handleClickOutside(event);
  }
  componentWillLoad() {
    this.hasDefaultTrigger = !this.element.querySelector('[slot="trigger"]');
  }
  disconnectedCallback() {
    var _a, _b;
    (_a = this.popper) === null || _a === void 0 ? void 0 : _a.destroy();
    (_b = this.tooltipRef) === null || _b === void 0 ? void 0 : _b.remove();
  }
  render() {
    const TriggerTag = this.tag;
    return (h(Host, null,
      h(TriggerTag, { "aria-describedby": this.idDescriber, class: getClassNames([
          'ld-tooltip__trigger',
          this.triggerType === 'click' && 'ld-tooltip__trigger--clickable',
        ]), onClick: this.handleToggleTrigger, onMouseEnter: this.handleShowTrigger, onFocus: this.handleShowTrigger, onMouseLeave: this.handleHideTrigger, onBlur: this.handleHideTrigger, part: "trigger focusable", ref: (element) => {
          this.triggerRef = element;
        }, type: "button" },
        h("ld-sr-only", null, "Info"),
        h("slot", { name: "trigger" },
          h("svg", { class: "ld-tooltip__icon", fill: "none", part: "icon", viewBox: "0 0 24 24" },
            h("path", { "clip-rule": "evenodd", d: "M12 23C18.0751 23 23 18.0751 23 12C23 5.9249 18.0751 1 12 1C5.9249 1 1 5.9249 1 12C1 18.0751 5.9249 23 12 23Z", "fill-rule": "evenodd", fill: "currentColor" }),
            h("path", { "clip-rule": "evenodd", d: "M11.9996 8.6477C12.9254 8.6477 13.6758 7.8973 13.6758 6.9715C13.6758 6.0458 12.9254 5.2953 11.9996 5.2953C11.0739 5.2953 10.3235 6.0458 10.3235 6.9715C10.3235 7.8973 11.0739 8.6477 11.9996 8.6477ZM10.8453 17.8038C11.1932 18.1517 11.6736 18.3256 12.2865 18.3256H13.4545C13.6864 18.3256 13.8023 18.2263 13.8023 18.0275V12.2873C13.8023 11.6744 13.6284 11.1939 13.2805 10.8461C12.9326 10.4982 12.4522 10.3242 11.8393 10.3242H10.6713C10.4394 10.3242 10.3235 10.4236 10.3235 10.6224V16.3626C10.3235 16.9755 10.4974 17.456 10.8453 17.8038Z", "fill-rule": "evenodd", 
              // TODO: replace color with icon-specific custom property and use ld-icon
              fill: "var(--ld-col-wht)" })))),
      h("ld-tooltip-popper", { "aria-hidden": this.visible ? undefined : 'true', arrow: this.arrow, hasDefaultTrigger: this.hasDefaultTrigger, id: this.idDescriber, part: "popper", size: this.size, ref: (element) => {
          this.tooltipRef = element;
        }, triggerType: this.triggerType },
        h("slot", null))));
  }
  static get is() { return "ld-tooltip"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["ld-tooltip.shadow.css"]
  }; }
  static get styleUrls() { return {
    "$": ["ld-tooltip.shadow.css"]
  }; }
  static get properties() { return {
    "arrow": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Show arrow"
      },
      "attribute": "arrow",
      "reflect": false
    },
    "disabled": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Disable tooltip trigger"
      },
      "attribute": "disabled",
      "reflect": false
    },
    "hideDelay": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Delay in ms until tooltip hides (only when trigger type is 'hover')"
      },
      "attribute": "hide-delay",
      "reflect": false,
      "defaultValue": "0"
    },
    "position": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "Position",
        "resolved": "\"bottom center\" | \"bottom left\" | \"bottom right\" | \"left bottom\" | \"left middle\" | \"left top\" | \"right bottom\" | \"right middle\" | \"right top\" | \"top center\" | \"top left\" | \"top right\"",
        "references": {
          "Position": {
            "location": "local"
          }
        }
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Position of the tooltip relative to the trigger element (also affects the arrow position)"
      },
      "attribute": "position",
      "reflect": false,
      "defaultValue": "'top center'"
    },
    "showDelay": {
      "type": "number",
      "mutable": false,
      "complexType": {
        "original": "number",
        "resolved": "number",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Delay in ms until tooltip shows (only when trigger type is 'hover')"
      },
      "attribute": "show-delay",
      "reflect": false,
      "defaultValue": "0"
    },
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'sm'",
        "resolved": "\"sm\"",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "The tooltip size (effects tooltip padding only)"
      },
      "attribute": "size",
      "reflect": false
    },
    "tag": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "The rendered HTML tag for the tooltip trigger."
      },
      "attribute": "tag",
      "reflect": false,
      "defaultValue": "'button'"
    },
    "tetherOptions": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "Partial<Tether.ITetherOptions> | string",
        "resolved": "string | { attachment?: string; bodyElement?: HTMLElement; classes?: { [className: string]: string | boolean; }; classPrefix?: string; constraints?: ITetherConstraint[]; element?: any; enabled?: boolean; offset?: string; optimizations?: any; target?: any; targetAttachment?: string; targetOffset?: string; targetModifier?: string; }",
        "references": {
          "Partial": {
            "location": "global"
          },
          "Tether": {
            "location": "global"
          }
        }
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Tether options object to be merged with the default options (optionally stringified)."
      },
      "attribute": "tether-options",
      "reflect": false
    },
    "triggerType": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'click' | 'hover'",
        "resolved": "\"click\" | \"hover\"",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Event type that triggers the tooltip"
      },
      "attribute": "trigger-type",
      "reflect": false,
      "defaultValue": "'hover'"
    }
  }; }
  static get states() { return {
    "hasDefaultTrigger": {},
    "visible": {}
  }; }
  static get methods() { return {
    "hideTooltip": {
      "complexType": {
        "signature": "() => Promise<void>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Hide tooltip",
        "tags": []
      }
    },
    "showTooltip": {
      "complexType": {
        "signature": "() => Promise<void>",
        "parameters": [],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Show tooltip",
        "tags": []
      }
    }
  }; }
  static get elementRef() { return "element"; }
  static get watchers() { return [{
      "propName": "disabled",
      "methodName": "updatePopper"
    }]; }
  static get listeners() { return [{
      "name": "click",
      "method": "handleClickOutside",
      "target": "window",
      "capture": false,
      "passive": false
    }, {
      "name": "touchend",
      "method": "handleTouchOutside",
      "target": "window",
      "capture": false,
      "passive": true
    }]; }
}
