// @emdgroup-liquid/liquid/dist/collection/components/ld-select/ld-select-popper/ld-select-popper.js
import { Component, Element, Event, h, Host, Method, Prop, State, Watch, } from '@stencil/core';
import { getClassNames } from '../../../utils/getClassNames';
/** @internal **/
export class LdSelectPopper {
  constructor() {
    /** Indicates if select element is expanded. */
    this.expanded = false;
    this.isPinned = false;
    this.shadowHeight = '100%';
    this.handleFilterInput = (ev) => {
      this.ldselectfilterchange.emit(ev.target.value);
    };
  }
  updatePinnedState() {
    this.isPinned = this.el.classList.contains('ld-tether-pinned');
  }
  updatePopperTheme(newValue, oldValue) {
    this.el.classList.remove(`ld-theme-${oldValue}`);
    if (newValue)
      this.el.classList.add(`ld-theme-${newValue}`);
  }
  /**
   * Focuses the tab
   */
  async updateShadowHeight(height) {
    this.shadowHeight = height;
  }
  componentWillLoad() {
    this.popperClass && this.el.classList.add(this.popperClass);
  }
  render() {
    return (h(Host, { style: {
        zIndex: this.isPinned ? '2147483647' : '2147483646',
        display: this.expanded ? 'block' : 'none',
      } },
      h("div", { class: getClassNames([
          'ld-select-popper',
          this.detached && 'ld-select-popper--detached',
          this.expanded && 'ld-select-popper--expanded',
          this.filter && 'ld-select-popper--filter',
          this.isPinned && 'ld-select-popper--pinned',
          this.size && `ld-select-popper--${this.size}`,
        ]), part: "popper" },
        this.filter && (h("input", { "aria-haspopup": "listbox", type: "text", placeholder: this.filterPlaceholder, class: "ld-select-popper__filter-input", part: "filter-input focusable", onInput: this.handleFilterInput })),
        h("div", { class: "ld-select-popper__scroll-container", part: "popper-scroll-container" },
          h("slot", null),
          h("div", { class: "ld-select-popper__shadow", style: { height: this.isPinned ? '100%' : this.shadowHeight }, part: "shadow" })))));
  }
  static get is() { return "ld-select-popper"; }
  static get encapsulation() { return "shadow"; }
  static get originalStyleUrls() { return {
    "$": ["ld-select-popper.shadow.css"]
  }; }
  static get styleUrls() { return {
    "$": ["ld-select-popper.shadow.css"]
  }; }
  static get properties() { return {
    "class": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "A watcher is applied to the CSS class in order to be able to react to tether changes."
      },
      "attribute": "class",
      "reflect": true
    },
    "detached": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Popper is visually detached from the select trigger element (there's a gap between the two)."
      },
      "attribute": "detached",
      "reflect": false
    },
    "expanded": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Indicates if select element is expanded."
      },
      "attribute": "expanded",
      "reflect": false,
      "defaultValue": "false"
    },
    "filter": {
      "type": "boolean",
      "mutable": false,
      "complexType": {
        "original": "boolean",
        "resolved": "boolean",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Set this property to `true` in order to enable an input field for filtering options."
      },
      "attribute": "filter",
      "reflect": false
    },
    "filterPlaceholder": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "The filter input placeholder."
      },
      "attribute": "filter-placeholder",
      "reflect": false
    },
    "popperClass": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Attaches CSS class to the select popper element."
      },
      "attribute": "popper-class",
      "reflect": false
    },
    "size": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "'sm' | 'lg'",
        "resolved": "\"lg\" | \"sm\"",
        "references": {}
      },
      "required": false,
      "optional": true,
      "docs": {
        "tags": [],
        "text": "Size of the select trigger button (required for applying the correct shadow height)."
      },
      "attribute": "size",
      "reflect": false
    },
    "theme": {
      "type": "string",
      "mutable": false,
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      },
      "required": false,
      "optional": false,
      "docs": {
        "tags": [],
        "text": "Since the select popper is located outside the select element, the theme needs to be applied as a prop."
      },
      "attribute": "theme",
      "reflect": false
    }
  }; }
  static get states() { return {
    "isPinned": {},
    "shadowHeight": {}
  }; }
  static get events() { return [{
      "method": "ldselectfilterchange",
      "name": "ldselectfilterchange",
      "bubbles": true,
      "cancelable": true,
      "composed": true,
      "docs": {
        "tags": [],
        "text": "Emitted on filter change with the filter input value."
      },
      "complexType": {
        "original": "string",
        "resolved": "string",
        "references": {}
      }
    }]; }
  static get methods() { return {
    "updateShadowHeight": {
      "complexType": {
        "signature": "(height: string) => Promise<void>",
        "parameters": [{
            "tags": [],
            "text": ""
          }],
        "references": {
          "Promise": {
            "location": "global"
          }
        },
        "return": "Promise<void>"
      },
      "docs": {
        "text": "Focuses the tab",
        "tags": []
      }
    }
  }; }
  static get elementRef() { return "el"; }
  static get watchers() { return [{
      "propName": "class",
      "methodName": "updatePinnedState"
    }, {
      "propName": "theme",
      "methodName": "updatePopperTheme"
    }]; }
}
