// @emdgroup-liquid/liquid/dist/cjs/ld-sr-live.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9117e53d.js');

const ldSrLiveShadowCss = "";

const LdSrLive = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
  }
  handleInfo(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.infoSuffix = this.infoSuffix ? '' : '​';
    this.info = event.detail + this.infoSuffix;
  }
  handleAlert(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.alertSuffix = this.alertSuffix ? '' : '​';
    this.alert = event.detail + this.alertSuffix;
  }
  render() {
    return (index.h("ld-sr-only", null, index.h("span", { role: "status", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.info), index.h("span", { role: "alert", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.alert)));
  }
};
LdSrLive.style = ldSrLiveShadowCss;

exports.ld_sr_live = LdSrLive;
