// @emdgroup-liquid/liquid/dist/cjs/ld-select-popper.cjs.entry.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const index = require('./index-9117e53d.js');
const getClassNames = require('./getClassNames-dd3906b3.js');

const ldSelectPopperShadowCss = ":host{--ld-select-popper-min-width:12.8125rem;--ld-select-popper-max-height:min(23.75rem,75vh - 1.25rem);--ld-select-popper-border-col:var(--ld-col-neutral-100);min-width:var(--ld-select-popper-min-width)}.ld-select-popper{min-width:100%}.ld-select-popper:not(.ld-select-popper--expanded){display:none}.ld-select-popper ::slotted(.ld-select__shadow){border-radius:var(--ld-br-m);box-shadow:var(--ld-shadow-sticky);display:block;height:100%;pointer-events:none;position:absolute;width:100%;z-index:-1}.ld-select-popper__scroll-container{border-bottom-left-radius:var(--ld-br-m);border-bottom-right-radius:var(--ld-br-m);border-top:solid var(--ld-select-popper-border-col) var(--ld-sp-1);max-height:var(--ld-select-popper-max-height);overflow-y:auto;overscroll-behavior:contain}.ld-select-popper--detached:not(.ld-select-popper--filter) .ld-select-popper__scroll-container,.ld-select-popper--pinned:not(.ld-select-popper--filter) .ld-select-popper__scroll-container{border-radius:var(--ld-br-m);border-top:0}.ld-select-popper__shadow{border-radius:var(--ld-br-m);bottom:0;box-shadow:var(--ld-shadow-sticky);height:calc(100% + var(--ld-select-min-height-md));pointer-events:none;position:absolute;width:100%;z-index:-1}.ld-select-popper--detached .ld-select-popper__shadow{height:100%!important}.ld-select-popper__filter-input{-webkit-appearance:none;appearance:none;background-color:var(--ld-col-wht);border:0;border-top:solid var(--ld-col-neutral-100) var(--ld-sp-1);box-sizing:border-box;color:var(--ld-col-neutral-900);font:var(--ld-typo-label-m);height:2.5rem;line-height:var(--ld-select-trigger-line-height);outline:none;padding:var(--ld-sp-8) var(--ld-sp-12);width:100%}.ld-select-popper__filter-input::placeholder{color:var(--ld-col-neutral-600)}.ld-select-popper--detached .ld-select-popper__filter-input,.ld-select-popper--pinned .ld-select-popper__filter-input{border-top:0;border-top-left-radius:var(--ld-br-m);border-top-right-radius:var(--ld-br-m)}";

const LdSelectPopper = class {
  constructor(hostRef) {
    index.registerInstance(this, hostRef);
    this.ldselectfilterchange = index.createEvent(this, "ldselectfilterchange", 7);
    /** Indicates if select element is expanded. */
    this.expanded = false;
    this.isPinned = false;
    this.shadowHeight = '100%';
    this.handleFilterInput = (ev) => {
      this.ldselectfilterchange.emit(ev.target.value);
    };
  }
  updatePinnedState() {
    this.isPinned = this.el.classList.contains('ld-tether-pinned');
  }
  updatePopperTheme(newValue, oldValue) {
    this.el.classList.remove(`ld-theme-${oldValue}`);
    if (newValue)
      this.el.classList.add(`ld-theme-${newValue}`);
  }
  /**
   * Focuses the tab
   */
  async updateShadowHeight(height) {
    this.shadowHeight = height;
  }
  componentWillLoad() {
    this.popperClass && this.el.classList.add(this.popperClass);
  }
  render() {
    return (index.h(index.Host, { style: {
        zIndex: this.isPinned ? '2147483647' : '2147483646',
        display: this.expanded ? 'block' : 'none',
      } }, index.h("div", { class: getClassNames.getClassNames([
        'ld-select-popper',
        this.detached && 'ld-select-popper--detached',
        this.expanded && 'ld-select-popper--expanded',
        this.filter && 'ld-select-popper--filter',
        this.isPinned && 'ld-select-popper--pinned',
        this.size && `ld-select-popper--${this.size}`,
      ]), part: "popper" }, this.filter && (index.h("input", { "aria-haspopup": "listbox", type: "text", placeholder: this.filterPlaceholder, class: "ld-select-popper__filter-input", part: "filter-input focusable", onInput: this.handleFilterInput })), index.h("div", { class: "ld-select-popper__scroll-container", part: "popper-scroll-container" }, index.h("slot", null), index.h("div", { class: "ld-select-popper__shadow", style: { height: this.isPinned ? '100%' : this.shadowHeight }, part: "shadow" })))));
  }
  get el() { return index.getElement(this); }
  static get watchers() { return {
    "class": ["updatePinnedState"],
    "theme": ["updatePopperTheme"]
  }; }
};
LdSelectPopper.style = ldSelectPopperShadowCss;

exports.ld_select_popper = LdSelectPopper;
