// @emdgroup-liquid/liquid/dist/components/ld-pagination.js
import { proxyCustomElement, HTMLElement, createEvent, h, Host } from '@stencil/core/internal/client';
import { g as getClassNames } from './getClassNames.js';
import { d as defineCustomElement$3 } from './ld-button2.js';
import { d as defineCustomElement$2 } from './ld-icon2.js';

const ldPaginationShadowCss = ".ld-pagination{--ld-pagination-animation-duration:1ms;--ld-pagination-sm-item-size:2rem;--ld-pagination-md-item-size:2.5rem;--ld-pagination-lg-item-size:3.125rem;--ld-pagination-column-size:var(--ld-pagination-default-column-size);--ld-pagination-default-column-size:calc(var(--ld-pagination-item-size) + var(--ld-pagination-items-space));--ld-pagination-dot-col:var(--ld-thm-primary);--ld-pagination-dots-indicator-size:calc(var(--ld-pagination-dots-size) + var(--ld-sp-8));--ld-pagination-dots-indicator-size-active:calc(var(--ld-pagination-dots-size) + var(--ld-sp-12));--ld-pagination-dots-sm-size:0.25rem;--ld-pagination-dots-md-size:0.375rem;--ld-pagination-dots-lg-size:0.5rem;--ld-pagination-dots-size:var(--ld-pagination-dots-md-size);--ld-pagination-dots-sm-space:var(--ld-sp-8);--ld-pagination-dots-md-space:var(--ld-sp-12);--ld-pagination-dots-lg-space:var(--ld-sp-16);--ld-pagination-dots-space:var(--ld-pagination-dots-md-space);--ld-pagination-item-size:var(--ld-pagination-md-item-size);--ld-pagination-items-space:var(--ld-sp-4);--ld-pagination-marker-col:var(--ld-thm-primary-alpha-lowest);--ld-pagination-more-indicator-font:var(--ld-typo-body-m);--ld-pagination-primary-col:var(--ld-thm-primary);--ld-pagination-selected-col:var(--thm-primary-active);align-items:center;display:inline-flex;list-style:none;margin:0;padding:var(--ld-sp-1);position:relative}.ld-pagination--sm{--ld-pagination-more-indicator-font:var(--ld-typo-body-s);--ld-pagination-item-size:var(--ld-pagination-sm-item-size);--ld-pagination-dots-size:var(--ld-pagination-dots-sm-size);--ld-pagination-dots-space:var(--ld-pagination-dots-sm-space)}.ld-pagination--lg{--ld-pagination-more-indicator-font:var(--ld-typo-body-l);--ld-pagination-item-size:var(--ld-pagination-lg-item-size);--ld-pagination-dots-size:var(--ld-pagination-dots-lg-size);--ld-pagination-dots-space:var(--ld-pagination-dots-lg-space)}.ld-pagination--brand-color{--ld-pagination-marker-col:var(--ld-col-wht-alpha-lowest);--ld-pagination-dot-col:var(--ld-col-wht);--ld-pagination-primary-col:var(--ld-col-wht);--ld-pagination-selected-col:var(--ld-col-wht)}.ld-pagination--brand-color ld-button::part(button){color:var(--ld-col-wht)}.ld-pagination--brand-color ld-button::part(button):hover{background-color:var(--ld-col-wht-alpha-low)}.ld-pagination__arrow:not(:last-child),.ld-pagination__sticky{margin-right:var(--ld-pagination-items-space)}.ld-pagination__item,.ld-pagination__sticky{width:var(--ld-pagination-item-size)}.ld-pagination__item ld-button::part(button),.ld-pagination__sticky ld-button::part(button){padding:0}.ld-pagination__item,.ld-pagination__more-indicator{opacity:0;transition:opacity .1s linear}.ld-pagination__item--visible,.ld-pagination__more-indicator--visible{opacity:1}.ld-pagination__more-indicator{color:var(--ld-pagination-primary-col);font:var(--ld-pagination-more-indicator-font);font-weight:700;height:var(--ld-pagination-item-size);line-height:1;pointer-events:none;position:relative}.ld-pagination__more-indicator--end{right:var(--ld-pagination-column-size)}.ld-pagination__more-indicator span{align-items:center;display:flex;height:100%;justify-content:center;position:absolute;white-space:nowrap;width:var(--ld-pagination-item-size)}.ld-pagination__slide-wrapper{height:var(--ld-pagination-item-size);position:relative;width:calc(var(--ld-pagination-slider-cols)*var(--ld-pagination-column-size))}.ld-pagination__slide-wrapper--transitioning{overflow:hidden}.ld-pagination__items{display:inline-flex;list-style:none;padding:0;pointer-events:none;position:absolute;transform:translateX(calc(var(--ld-pagination-slide-index)*var(--ld-pagination-column-size)*-1));transition:transform .1s ease}.ld-pagination__item{left:calc(var(--ld-pagination-item-pos)*var(--ld-pagination-column-size));position:absolute}.ld-pagination__item--visible{pointer-events:auto}.ld-pagination__item--selected ld-button{--ld-button-bg-col:#0000}.ld-pagination__item--selected ld-button::part(button){color:var(--ld-pagination-selected-col);font-weight:900}.ld-pagination__item--selected ld-button .ld-pagination__dot{background-color:var(--ld-pagination-primary-col)}.ld-pagination__marker{background:var(--ld-pagination-marker-col);border-radius:var(--ld-br-m);height:var(--ld-pagination-item-size);position:absolute;transform:translateX(calc(var(--ld-pagination-column-size)*var(--ld-pagination-selected-index)));transition:transform var(--ld-pagination-animation-duration) ease,opacity var(--ld-pagination-animation-duration) ease;width:var(--ld-pagination-item-size);z-index:-1}@media (prefers-reduced-motion:no-preference){.ld-pagination__marker{--ld-pagination-animation-duration:0.1s}}.ld-pagination__marker--hidden{opacity:0}.ld-pagination--dots{--ld-pagination-item-size:var(--ld-pagination-dots-size);--ld-pagination-items-space:var(--ld-pagination-dots-space);--ld-pagination-column-size:max(var(--ld-pagination-default-column-size),var(--ld-pagination-dots-indicator-size))}.ld-pagination--dots .ld-pagination__arrow:not(:last-child),.ld-pagination--dots .ld-pagination__sticky{margin-right:var(--ld-sp-4)}.ld-pagination--dots .ld-pagination__slide-wrapper{height:var(--ld-pagination-column-size);margin-right:var(--ld-sp-4)}.ld-pagination--dots .ld-pagination__item{align-items:center;display:flex;height:var(--ld-pagination-column-size);justify-content:center;left:calc(var(--ld-pagination-item-pos)*var(--ld-pagination-column-size));width:var(--ld-pagination-column-size)}.ld-pagination--dots .ld-pagination__item:not(.ld-pagination__item--selected) .ld-pagination__dot:after{background-color:var(--ld-pagination-primary-col);opacity:.3}.ld-pagination__dot{align-items:center;display:flex;height:var(--ld-pagination-column-size);justify-content:center;width:var(--ld-pagination-column-size)}.ld-pagination__dot::part(button){align-items:center;border-radius:var(--ld-br-full);display:flex;height:var(--ld-pagination-dots-indicator-size);justify-content:center;min-height:auto;min-width:auto;position:absolute;width:var(--ld-pagination-dots-indicator-size)}.ld-pagination__dot::part(button):active{height:var(--ld-pagination-dots-indicator-size-active);width:var(--ld-pagination-dots-indicator-size-active)}.ld-pagination__dot::part(button):focus{outline-color:var(--ld-thm-primary);outline-offset:calc(var(--ld-sp-1)*-1)}.ld-pagination__dot::part(button):focus:not(:focus-visible){outline:none}.ld-pagination__dot:after{background-color:var(--ld-pagination-dot-col);border-radius:50%;content:\"\";height:var(--ld-pagination-dots-size);pointer-events:none;position:absolute;width:var(--ld-pagination-dots-size)}";

const BUFFER_SIZE = 20;
const LdPagination$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
    this.ldchange = createEvent(this, "ldchange", 7);
    /** Hide the buttons to navigate forward/backward. */
    this.hidePrevNext = false;
    /** Hide the buttons to navigate to the first/last item. */
    this.hideStartEnd = false;
    /** Label to communicate the type of an item. */
    this.itemLabel = 'Page';
    /** The number of items/pages available for pagination (required to let the user jump to the last item/page). */
    this.length = Infinity;
    /** Items display mode, default as numbers. */
    this.mode = 'numbers';
    /** Number of next/previous items visible. */
    this.offset = 2;
    /** The currently selected item (an index of `-1` means nothing is selected). */
    this.selectedIndex = 0;
    /** Number of items permanently visible at the start/end. */
    this.sticky = 0;
    this.maxSliderColumns = 0;
    this.renderMoreIndicators = false;
    this.renderSticky = false;
    this.slidableItems = [];
    this.sliderContent = [];
    this.transitioning = false;
    this.visibleItemsInSlider = 0;
    this.handleTransitionEnd = () => {
      this.transitioning = false;
    };
    this.handleTransitionStart = () => {
      this.transitioning = true;
    };
    // pageNumber is 1-based
    this.renderItem = (itemNumber, showFrom, showTo) => {
      const isDots = this.mode === 'dots';
      const isHidden = (this.renderMoreIndicators || isDots) &&
        (itemNumber < showFrom || itemNumber > showTo);
      const isSelected = itemNumber === this.selectedIndex + 1;
      return (h("li", { "aria-hidden": isHidden ? 'true' : undefined, class: getClassNames([
          'ld-pagination__item',
          !isHidden && 'ld-pagination__item--visible',
          isSelected && 'ld-pagination__item--selected',
        ]), key: itemNumber, style: { '--ld-pagination-item-pos': `${itemNumber - 1}` } }, h("ld-button", { "aria-current": isSelected ? 'true' : undefined, "aria-label": isHidden ? undefined : `${this.itemLabel} ${itemNumber}`, class: isDots ? 'ld-pagination__dot' : undefined, "ld-tabindex": isHidden ? -1 : undefined, mode: "ghost", onClick: () => {
          this.selectedIndex = itemNumber - 1;
        }, part: "item focusable", size: this.size }, !isDots && itemNumber)));
    };
    this.calculateSliderContent = () => {
      const directlyReachableFirstItems = this.maxSliderColumns + this.sticky - 1;
      const directlyReachableLastItems = this.length - this.maxSliderColumns - this.sticky + 1;
      this.sliderContent = this.slidableItems.filter((itemNumber) => (itemNumber > this.selectedIndex - BUFFER_SIZE &&
        itemNumber <= this.selectedIndex + BUFFER_SIZE) ||
        // render the first/last items in case the user can navigate there directly
        ((!this.hideStartEnd || this.sticky > 0) &&
          (itemNumber <= directlyReachableFirstItems ||
            itemNumber > directlyReachableLastItems)));
    };
  }
  handleSelectedIndex() {
    if (this.selectedIndex < -1) {
      this.selectedIndex = -1;
    }
    else if (this.selectedIndex >= this.length) {
      this.selectedIndex = this.length - 1;
    }
    else {
      this.ldchange.emit(this.selectedIndex);
    }
  }
  componentWillLoad() {
    this.visibleItemsInSlider = this.offset * 2 + 1;
    const maxVisibleItems = this.sticky * 2 + this.visibleItemsInSlider;
    this.maxSliderColumns = this.visibleItemsInSlider + 2;
    this.renderSticky = this.sticky > 0 && this.mode !== 'dots';
    this.renderMoreIndicators =
      this.mode !== 'dots' && this.length > maxVisibleItems + 2;
    this.slidableItems = Array.from({
      length: this.length === Infinity ? 9999 : this.length - this.sticky * 2,
    }).map((_, index) => index + this.sticky + 1);
    if (this.length < 1) {
      this.length = 1;
    }
    if (this.selectedIndex < -1) {
      this.selectedIndex = -1;
    }
    else if (this.selectedIndex >= this.length) {
      this.selectedIndex = this.length - 1;
    }
    this.calculateSliderContent();
  }
  componentDidLoad() {
    if (this.wrapperRef) {
      this.wrapperRef.addEventListener('transitionstart', this.handleTransitionStart);
      this.wrapperRef.addEventListener('transitionend', this.handleTransitionEnd);
    }
  }
  disconnectedCallback() {
    if (this.wrapperRef) {
      this.wrapperRef.removeEventListener('transitionstart', this.handleTransitionStart);
      this.wrapperRef.removeEventListener('transitionend', this.handleTransitionEnd);
    }
  }
  render() {
    const isDots = this.mode === 'dots';
    const styleDots = isDots && this.space
      ? {
        '--ld-pagination-dots-space': this.space === '0' ? '0px' : this.space,
      }
      : undefined;
    // +1 because it must be the index right to the centered item
    const showStartMoreIndicator = this.renderMoreIndicators &&
      this.selectedIndex > this.sticky + this.offset + 1;
    // -1 because it is 0-based and another -1 because it must be the index left to the centered item
    const showEndMoreIndicator = this.renderMoreIndicators &&
      this.selectedIndex < this.length - this.offset - this.sticky - 2;
    const showFrom = 
    // +1 because it is not 0-based
    Math.max(Math.min(this.selectedIndex -
      this.offset -
      // start hiding numbers
      (showStartMoreIndicator ? 0 : 1), this.length -
      this.visibleItemsInSlider -
      this.sticky -
      1 +
      (isDots ? -1 : 0)), this.sticky) + 1;
    // +1 because it is not 0-based
    const showTo = Math.min(Math.max(this.selectedIndex + (showEndMoreIndicator ? 0 : 1), this.offset + this.sticky + 1 + (isDots ? 1 : 0)) + this.offset, this.length - this.sticky) + 1;
    return (h(Host, { role: "navigation" }, h("ul", { "aria-label": "Pagination", class: getClassNames([
        'ld-pagination',
        this.size && `ld-pagination--${this.size}`,
        isDots && `ld-pagination--dots`,
        this.brandColor && 'ld-pagination--brand-color',
      ]), part: "wrapper", style: styleDots }, !this.hideStartEnd && (h("li", { class: "ld-pagination__arrow" }, h("ld-button", { "aria-label": this.startLabel
        ? undefined
        : `First ${this.itemLabel.toLocaleLowerCase()}`, disabled: this.selectedIndex < 1 ? true : undefined, mode: "ghost", onClick: () => {
        this.selectedIndex = 0;
      }, part: "arrow start focusable", size: this.size }, this.startLabel ? (this.startLabel) : (h("ld-icon", { name: "arrow-double-left", size: this.size }))))), !this.hidePrevNext && (h("li", { class: "ld-pagination__arrow" }, h("ld-button", { "aria-label": this.prevLabel
        ? undefined
        : `Previous ${this.itemLabel.toLocaleLowerCase()}`, disabled: this.selectedIndex < 1, mode: "ghost", onClick: () => {
        this.selectedIndex -= 1;
      }, part: "arrow prev focusable", size: this.size }, this.prevLabel ? (this.prevLabel) : (h("ld-icon", { name: "arrow-left", size: this.size }))))), this.renderSticky &&
      Array.from({ length: Math.min(this.sticky, this.length) }).map((_, index) => {
        return (h("li", { class: "ld-pagination__sticky" }, h("ld-button", { "aria-label": `${this.itemLabel} ${index + 1}`, mode: "ghost", onClick: () => {
            this.selectedIndex = index;
          }, part: "sticky item focusable", size: this.size }, index + 1)));
      }), this.renderMoreIndicators && (h("li", { class: getClassNames([
        'ld-pagination__more-indicator',
        showStartMoreIndicator &&
          'ld-pagination__more-indicator--visible',
      ]), part: "more-indicator" }, h("span", null, ". . ."))), h("li", { class: getClassNames([
        'ld-pagination__slide-wrapper',
        this.transitioning &&
          'ld-pagination__slide-wrapper--transitioning',
      ]), part: "slide-wrapper", ref: (ref) => (this.wrapperRef = ref), style: {
        '--ld-pagination-slider-cols': `${Math.min(this.slidableItems.length, this.maxSliderColumns)}`,
      } }, h("ul", { class: "ld-pagination__items", part: "items", style: {
        '--ld-pagination-slide-index': `${Math.max(Math.min(this.selectedIndex - this.offset - 1, 
        // -1 because of the "..." and another -1 because it is 0-based
        this.length - this.visibleItemsInSlider - this.sticky - 2), 
        // marker position in case of length < sticky
        Math.min(this.sticky, this.length), 0)}`,
      } }, !isDots && (h("li", { class: getClassNames([
        'ld-pagination__marker',
        this.selectedIndex < 0 && 'ld-pagination__marker--hidden',
      ]), key: "marker", onTransitionEnd: this.calculateSliderContent, part: "marker", style: {
        '--ld-pagination-selected-index': `${Math.max(this.selectedIndex, 0)}`,
      } })), this.length > 0 &&
      this.sliderContent.map((itemNumber) => this.renderItem(itemNumber, showFrom, showTo)))), this.renderMoreIndicators && (h("li", { class: getClassNames([
        'ld-pagination__more-indicator ld-pagination__more-indicator--end',
        showEndMoreIndicator &&
          'ld-pagination__more-indicator--visible',
      ]), part: "more-indicator" }, h("span", null, ". . ."))), this.renderSticky &&
      this.length < Infinity &&
      Array.from({ length: this.sticky })
        .map((_, index) => {
        const itemNumber = this.length - index;
        if (itemNumber <= this.sticky) {
          return null;
        }
        return (h("li", { class: "ld-pagination__sticky" }, h("ld-button", { mode: "ghost", onClick: () => {
            this.selectedIndex = itemNumber - 1;
          }, part: "sticky item focusable", size: this.size }, itemNumber)));
      })
        .reverse(), !this.hidePrevNext && (h("li", { class: "ld-pagination__arrow" }, h("ld-button", { "aria-label": this.nextLabel
        ? undefined
        : `Next ${this.itemLabel.toLocaleLowerCase()}`, disabled: this.selectedIndex >= this.length - 1, mode: "ghost", onClick: () => {
        this.selectedIndex += 1;
      }, part: "arrow next focusable", size: this.size }, this.nextLabel ? (this.nextLabel) : (h("ld-icon", { name: "arrow-right", size: this.size }))))), this.length < Infinity && !this.hideStartEnd && (h("li", { class: "ld-pagination__arrow" }, h("ld-button", { "aria-label": this.endLabel
        ? undefined
        : `Last ${this.itemLabel.toLocaleLowerCase()}`, disabled: this.selectedIndex >= this.length - 1, mode: "ghost", onClick: () => {
        this.selectedIndex = this.length - 1;
      }, part: "arrow end focusable", size: this.size }, this.endLabel ? (this.endLabel) : (h("ld-icon", { name: "arrow-double-right", size: this.size }))))))));
  }
  static get assetsDirs() { return ["assets"]; }
  static get watchers() { return {
    "selectedIndex": ["handleSelectedIndex"],
    "length": ["componentWillLoad"],
    "mode": ["componentWillLoad"],
    "offset": ["componentWillLoad"],
    "sticky": ["componentWillLoad"]
  }; }
  static get style() { return ldPaginationShadowCss; }
}, [1, "ld-pagination", {
    "brandColor": [4, "brand-color"],
    "endLabel": [1, "end-label"],
    "hidePrevNext": [4, "hide-prev-next"],
    "hideStartEnd": [4, "hide-start-end"],
    "itemLabel": [1, "item-label"],
    "length": [1026],
    "mode": [1],
    "nextLabel": [1, "next-label"],
    "offset": [2],
    "prevLabel": [1, "prev-label"],
    "selectedIndex": [1026, "selected-index"],
    "size": [1],
    "space": [1],
    "startLabel": [1, "start-label"],
    "sticky": [2],
    "maxSliderColumns": [32],
    "renderMoreIndicators": [32],
    "renderSticky": [32],
    "slidableItems": [32],
    "sliderContent": [32],
    "transitioning": [32],
    "visibleItemsInSlider": [32]
  }]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["ld-pagination", "ld-button", "ld-icon"];
  components.forEach(tagName => { switch (tagName) {
    case "ld-pagination":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, LdPagination$1);
      }
      break;
    case "ld-button":
      if (!customElements.get(tagName)) {
        defineCustomElement$3();
      }
      break;
    case "ld-icon":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}
defineCustomElement$1();

const LdPagination = LdPagination$1;
const defineCustomElement = defineCustomElement$1;

export { LdPagination, defineCustomElement };
