// @emdgroup-liquid/liquid/dist/components/ld-sr-live.js
import { proxyCustomElement, HTMLElement, h } from '@stencil/core/internal/client';
import { d as defineCustomElement$2 } from './ld-sr-only2.js';

const ldSrLiveShadowCss = "";

const LdSrLive$1 = /*@__PURE__*/ proxyCustomElement(class extends HTMLElement {
  constructor() {
    super();
    this.__registerHost();
    this.__attachShadow();
  }
  handleInfo(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.infoSuffix = this.infoSuffix ? '' : '​';
    this.info = event.detail + this.infoSuffix;
  }
  handleAlert(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.alertSuffix = this.alertSuffix ? '' : '​';
    this.alert = event.detail + this.alertSuffix;
  }
  render() {
    return (h("ld-sr-only", null, h("span", { role: "status", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.info), h("span", { role: "alert", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.alert)));
  }
  static get style() { return ldSrLiveShadowCss; }
}, [1, "ld-sr-live", {
    "info": [32],
    "alert": [32]
  }, [[9, "ldSrLiveInfo", "handleInfo"], [9, "ldSrLiveAlert", "handleAlert"]]]);
function defineCustomElement$1() {
  if (typeof customElements === "undefined") {
    return;
  }
  const components = ["ld-sr-live", "ld-sr-only"];
  components.forEach(tagName => { switch (tagName) {
    case "ld-sr-live":
      if (!customElements.get(tagName)) {
        customElements.define(tagName, LdSrLive$1);
      }
      break;
    case "ld-sr-only":
      if (!customElements.get(tagName)) {
        defineCustomElement$2();
      }
      break;
  } });
}
defineCustomElement$1();

const LdSrLive = LdSrLive$1;
const defineCustomElement = defineCustomElement$1;

export { LdSrLive, defineCustomElement };
