// @emdgroup-liquid/liquid/dist/esm/ld-breadcrumbs.entry.js
import { r as registerInstance, h, g as getElement } from './index-51c17921.js';

const ldBreadcrumbsCss = ".ld-breadcrumbs,:host{--ld-crumb-icon-gap:0.5em;--ld-crumb-gap:0.6em;line-height:1.5}.ld-breadcrumbs>ol{display:flex;flex-wrap:wrap;list-style:none;margin:0;padding:0}";

const LdBreadcrumbs = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
    this.updateCurrent = () => {
      const crumbs = this.el.querySelectorAll('ld-crumb');
      crumbs.forEach((crumb) => {
        crumb.current = undefined;
      });
      crumbs[crumbs.length - 1].current = true;
    };
  }
  componentDidLoad() {
    this.observer = new MutationObserver(this.updateCurrent);
    this.observer.observe(this.el, {
      subtree: true,
      childList: true,
      attributes: false,
    });
    this.updateCurrent();
  }
  disconnectedCallback() {
    if (this.observer)
      this.observer.disconnect();
  }
  render() {
    return (h("nav", { "aria-label": "Breadcrumbs", class: "ld-breadcrumbs" }, h("ol", { class: "ld-breadcrumbs__list", part: "list" }, h("slot", null))));
  }
  get el() { return getElement(this); }
};
LdBreadcrumbs.style = ldBreadcrumbsCss;

export { LdBreadcrumbs as ld_breadcrumbs };
