// @emdgroup-liquid/liquid/dist/esm/ld-sr-live.entry.js
import { r as registerInstance, h } from './index-51c17921.js';

const ldSrLiveShadowCss = "";

const LdSrLive = class {
  constructor(hostRef) {
    registerInstance(this, hostRef);
  }
  handleInfo(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.infoSuffix = this.infoSuffix ? '' : '​';
    this.info = event.detail + this.infoSuffix;
  }
  handleAlert(event) {
    // MAGIC: the second “pseudo-empty” string contains a zero width space, which is required
    // in order to repeat an alert message, without altering the printed message displayed on the screen.
    this.alertSuffix = this.alertSuffix ? '' : '​';
    this.alert = event.detail + this.alertSuffix;
  }
  render() {
    return (h("ld-sr-only", null, h("span", { role: "status", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.info), h("span", { role: "alert", "aria-live": "polite", "aria-relevant": "all", "aria-atomic": "true" }, this.alert)));
  }
};
LdSrLive.style = ldSrLiveShadowCss;

export { LdSrLive as ld_sr_live };
